lock-api/src/test/java/com/palantir/lock/client/ProfilingTimelockServiceTest.java

```diff
@@ -0,0 +1,236 @@
+
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        assertThatThrownBy(profilingTimelockService::getFreshTimestamp).isEqualTo(exception);
+
+        verify(delegate).getFreshTimestamp();
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+    }
+
+    @Test
+    public void doesNotLogIfLockIsSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.lock(LockRequest.of(ImmutableSet.of(StringLockDescriptor.of("exclusive")), 1234));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void doesNotLogIfWaitForLocksIsSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.waitForLocks(WaitForLocksRequest.of(
+                ImmutableSet.of(StringLockDescriptor.of("combination")), 123));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void logsSlowestOperationIfMultipleOperationsExceedTheSlowThresholdIfFirst() {
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.startIdentifiedAtlasDbTransaction();
+
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", TWO_CENTURIES);
+    }
+
+    @Test
+    public void logsSlowestOperationIfMultipleOperationsExceedTheSlowThresholdIfNotFirst() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        allowLogging();
+        profilingTimelockService.startIdentifiedAtlasDbTransaction();
+
+        verifyLoggerInvokedWithSpecificProfile("startIdentifiedAtlasDbTransaction", TWO_CENTURIES);
+    }
+
+    @Test
+    public void logsTheSlowestNonBannedOperationAboveThreshold() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        allowLogging();
+        profilingTimelockService.lock(LockRequest.of(ImmutableSet.of(StringLockDescriptor.of("mortice")), 2366));
+
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+    }
+
+    @Test
+    public void logsOnlyOnceEvenIfManyRequestsAreSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        for (int i = 0; i < 100; i++) {
+            profilingTimelockService.lockImmutableTimestamp();
+        }
+
+        makeOperationsTakeSpecifiedDuration(SHORT_DURATION);
+        allowLogging();
+        profilingTimelockService.startIdentifiedAtlasDbTransaction();
+
+        verifyLoggerInvokedWithSpecificProfile("lockImmutableTimestamp", LONG_DURATION);
```

Comment:
have to verify that it was only done once?, I can't remember if calling `verify` by is saying at least once or exactly once.

Reply:
Yep - `verify(foo).bar()` unqualified means exactly once.